hook/wrapper.go

```diff
@@ -169,110 +170,94 @@ func NewScriptWrapper(opts ...scriptWrapperOpt) (*ScriptWrapper, error) {
-	wrap.afterEnvFile, err = shell.TempFileWithExtension(
-		"buildkite-agent-bootstrap-hook-env-after",
-	)
+	wrap.afterEnvPath, err = tmpfile.KeepExtensionAndClose(hookWrapperDir, "hook-after-env")
 	if err != nil {
 		return nil, err
 	}
-	if err := wrap.afterEnvFile.Close(); err != nil {
-		return nil, fmt.Errorf("failed to close after env file: %w", err)
-	}
 
 	absolutePathToHook, err := filepath.Abs(wrap.hookPath)
 	if err != nil {
 		return nil, fmt.Errorf("finding absolute path to %q: %w", wrap.hookPath, err)
 	}
 
-	tmplInput := scriptTemplateInput{
-		ShebangLine:       shebang,
-		BeforeEnvFileName: wrap.beforeEnvFile.Name(),
-		AfterEnvFileName:  wrap.afterEnvFile.Name(),
-		PathToHook:        absolutePathToHook,
-	}
-
-	// Create the hook runner code
-	buf := &strings.Builder{}
-	switch {
-	case isWindows && !isPOSIXHook && !isPwshHook:
-		batchScriptTmpl.Execute(buf, tmplInput)
-
-	case isWindows && isPwshHook:
-		powershellScriptTmpl.Execute(buf, tmplInput)
-
-	default:
-		posixShellScriptTmpl.Execute(buf, tmplInput)
-	}
-	script := buf.String()
-
-	// Write the hook script to the runner then close the file so we can run it
-	_, err = wrap.scriptFile.WriteString(script)
-	if err != nil {
+	if wrap.wrapperPath, err = WriteHookWrapper(
+		tmpl,
+		WrapperTemplateInput{
+			ShebangLine:       shebang,
+			BeforeEnvFileName: wrap.beforeEnvPath,
+			AfterEnvFileName:  wrap.afterEnvPath,
+			PathToHook:        absolutePathToHook,
+		},
+		scriptFileName,
+	); err != nil {
 		return nil, err
 	}
 
-	// Make script executable
-	err = utils.ChmodExecutable(wrap.scriptFile.Name())
-	if err != nil {
-		return nil, err
-	}
+	return wrap, nil
+}
 
-	// the defered close attempt will discard any errors,
-	// and attempting to execute a unclosed script will fail with ETXTBSY
-	if err := wrap.scriptFile.Close(); err != nil {
-		return nil, fmt.Errorf("failed to close script file: %w", err)
+func WriteHookWrapper(
+	tmpl *template.Template,
+	input WrapperTemplateInput,
+	hookWrapperName string,
+) (name string, err error) {
+	f, err := tmpfile.KeepExtensionWithMode(hookWrapperDir, hookWrapperName, 0o700)
+	if err != nil {
+		return "", err
 	}
-	return wrap, nil
+	defer func() {
+		if cerr := f.Close(); cerr != nil {
+			err = fmt.Errorf("closing file: %w, other: %w", cerr, err)
+		}
+	}()
+	return f.Name(), tmpl.Execute(f, input)
 }
 
 // Path returns the path to the wrapper script, this is the one that should be executed
-func (wrap *ScriptWrapper) Path() string {
-	return wrap.scriptFile.Name()
+func (w *Wrapper) Path() string {
+	return w.wrapperPath
 }
 
-// Close cleans up the wrapper script and the environment files
-func (wrap *ScriptWrapper) Close() {
-	os.Remove(wrap.scriptFile.Name())
-	os.Remove(wrap.beforeEnvFile.Name())
-	os.Remove(wrap.afterEnvFile.Name())
+// Close cleans up the wrapper script and the environment files. Ignores errors, in
+// particular the error from os.Remove if the file doesn't exist.
```

Comment:
Do we know why it ignored errors before? just a "best effort" cleanup?

Reply:
IDK, I belive we started to remove them here in d062287f7a7074ccb3d269f3994a4a437283e811, but that looks like a marathon commit so I have no I idea why.

I think it might be a good idea to put all these in the same directory and clean the entire directory. Happy to do that here. I can change it to pass up the error then too. LMK what you think.

Reply:
Oh dear, history homework. This is fine.